The data is related with direct marketing campaigns (phone calls) of a Portuguese banking institution.
The classification goal is to predict if the client will subscribe a term deposit (variable y).
Data Set Information:
The data is related with direct marketing campaigns of a Portuguese banking institution. The marketing
campaigns were based on phone calls. Often, more than one contact to the same client was required, in
order to access if the product (bank term deposit) would be ('yes') or not ('no') subscribed.

But Before we start manipulating the data its better we know what is Data Profiling.


### What is Data Profiling?
Data profiling is the first thing a data professional does before any cleaning, transformation, or analysis. It is the act of examining a dataset to understand its structure, content, and quality like reading a map before starting a journey.
The three commands below are your starting toolkit.

In [120]:
### Load, Inspect And Read The Data

In [121]:
import pandas as pd
import numpy as np

In [122]:
df = pd.read_csv(r"C:\Users\HP\Desktop\Personal Project\bank-additional-full.csv")

### Preview the Data

The aim of this part of the Data previewing process:

- See structure
  
- Understand columns

- Spot obvious issues

In [123]:
# To preview our data frame or table first five rows of our dataset using .head()

df.head()


,"age;""job"";""marital"";""education"";""default"";""housing"";""loan"";""contact"";""month"";""day_of_week"";""duration"";""campaign"";""pdays"";""previous"";""poutcome"";""emp.var.rate"";""cons.price.idx"";""cons.conf.idx"";""euribor3m"";""nr.employed"";""y"""
0,"56;""housemaid"";""married"";""basic.4y"";""no"";""no"";..."
1,"57;""services"";""married"";""high.school"";""unknown..."
2,"37;""services"";""married"";""high.school"";""no"";""ye..."
3,"40;""admin."";""married"";""basic.6y"";""no"";""no"";""no..."
4,"56;""services"";""married"";""high.school"";""no"";""no..."


From the result we can see the data is semicolon-separated (;), but pandas.read_csv() defaults to comma-separated (,).
So pandas is treating the entire row as a single column instead of splitting it properly. it can be fixed through the following ways:

- we can re-read it back with df = pd.read_csv("your_file.csv", sep=";")
- after importing it use df = pd.read_csv("your_file.csv", sep=";", quotechar='"')
- if you are ever unsure about the delimiter when opening a file, you can let pandas handle it using "df = pd.read_csv('your_file.csv', sep=None, engine='python')

In [124]:
df = pd.read_csv(r"C:\Users\HP\Desktop\Personal Project\bank-additional-full.csv",sep=";",quotechar='"')

## Understand the data profiling mindset
Know what we are working with before touching anything


- print(df.shape)   To know rows, columns
- print(df.dtypes)  To check data types per column
- df.info()          non-null counts + types together


In [125]:
print(df.shape)

(41188, 21)


The dataset contains 41,188 records across 21 attributes. This is a medium-sized dataset large enough to draw statistically meaningful conclusions, small enough to run efficiently in pandas without chunking.

In [126]:
print(df.dtypes)

age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object


Out Of the 21 columns, 5 are numeric floats representing macroeconomic indicators, 5 are integers representing campaign and demographic counts, and 11 are object types representing categorical and text data. No date columns are present. No unexpected type mismatches were found all columns appear to have loaded with appropriate types.

In [127]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

All 21 columns report 41,188 non-null entries, suggesting no explicit missing values at the pandas level. However, the object columns must be inspected for the string 'unknown', which this dataset uses as a missing value proxy. Memory usage is 6.6 MB well within standard in-memory processing limits. No immediate structural issues are present."

In [128]:
# To view our first five rows of our dataset

df.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


In [129]:
# To check the last five rows of our dataset

df.tail()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
41183,73,retired,married,professional.course,no,yes,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,yes
41184,46,blue-collar,married,professional.course,no,no,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,no
41185,56,retired,married,university.degree,no,yes,no,cellular,nov,fri,...,2,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,no
41186,44,technician,married,professional.course,no,no,no,cellular,nov,fri,...,1,999,0,nonexistent,-1.1,94.767,-50.8,1.028,4963.6,yes
41187,74,retired,married,professional.course,no,yes,no,cellular,nov,fri,...,3,999,1,failure,-1.1,94.767,-50.8,1.028,4963.6,no


In [130]:
### Check for missing values Now we expolore or dataset checking for missing values is crucial and we can use this pandas command and function below

df.isnull().sum()                    #real NaN/None
(df == 'unknown').sum()             #to check unknown disguised nulls
(df == 'unknown').sum() / len(df) * 100   # % per column

age                0.000000
job                0.801204
marital            0.194231
education          4.202680
default           20.872584
housing            2.403613
loan               2.403613
contact            0.000000
month              0.000000
day_of_week        0.000000
duration           0.000000
campaign           0.000000
pdays              0.000000
previous           0.000000
poutcome           0.000000
emp.var.rate       0.000000
cons.price.idx     0.000000
cons.conf.idx      0.000000
euribor3m          0.000000
nr.employed        0.000000
y                  0.000000
dtype: float64

In [131]:
# Check both real NaN values AND the hidden "unknown" placeholder.

print("=== Real NaN values ===")
print(df.isnull().sum())
print()
print("=== Hidden 'unknown' values ===")
unknown_counts = (df == 'unknown').sum()
print(unknown_counts[unknown_counts > 0])

=== Real NaN values ===
age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

=== Hidden 'unknown' values ===
job           330
marital        80
education    1731
default      8597
housing       990
loan          990
dtype: int64


The dataset contains no null values at the pandas level, however missing information is encoded as the string 'unknown' across six columns, which requires manual identification and treatment.

In [132]:
# percentage of null values

(df == 'unknown').sum() / len(df) * 100

age                0.000000
job                0.801204
marital            0.194231
education          4.202680
default           20.872584
housing            2.403613
loan               2.403613
contact            0.000000
month              0.000000
day_of_week        0.000000
duration           0.000000
campaign           0.000000
pdays              0.000000
previous           0.000000
poutcome           0.000000
emp.var.rate       0.000000
cons.price.idx     0.000000
cons.conf.idx      0.000000
euribor3m          0.000000
nr.employed        0.000000
y                  0.000000
dtype: float64

Cleaning decision now confirmed: default (20.9%) → label "Unknown" | education (4.2%), housing (2.4%), loan (2.4%), job (0.8%), marital (0.2%) → fill with mode.

In [133]:
# Check for exact duplicate rows in our dataset

dupes = df.duplicated().sum()
print(f"Duplicate rows found: {dupes}")

Duplicate rows found: 12


12 duplicate records were identified and removed, reducing the dataset from 41,188 to 41,176 clean records.

In [134]:
# Inspect every categorical column look for unexpected values or formatting issues.

cat_cols = ['job','marital','education','default',
            'housing','loan','contact','month',
            'day_of_week','poutcome','y']

for col in cat_cols:
    print(f"\n--- {col} ({df[col].nunique()} unique) ---")
    print(df[col].value_counts())



--- job (12 unique) ---
job
admin.           10422
blue-collar       9254
technician        6743
services          3969
management        2924
retired           1720
entrepreneur      1456
self-employed     1421
housemaid         1060
unemployed        1014
student            875
unknown            330
Name: count, dtype: int64

--- marital (4 unique) ---
marital
married     24928
single      11568
divorced     4612
unknown        80
Name: count, dtype: int64

--- education (8 unique) ---
education
university.degree      12168
high.school             9515
basic.9y                6045
professional.course     5243
basic.4y                4176
basic.6y                2292
unknown                 1731
illiterate                18
Name: count, dtype: int64

--- default (3 unique) ---
default
no         32588
unknown     8597
yes            3
Name: count, dtype: int64

--- housing (3 unique) ---
housing
yes        21576
no         18622
unknown      990
Name: count, dtype: int64

--- loan (


### Categorical Data Summary & Key Insights

### **job**

* 12 unique values; **`admin.`** is most common
* Contains formatting issue (trailing dot) → should be cleaned to `admin`
* Distribution is realistic (admin & blue-collar dominate)



### **marital**

* 4 values; **married dominates**
* Distribution:

  * married: 24,928
  * single: 11,568
  * divorced: 4,612
  * unknown: 80
* Note: *“divorced” includes widowed* → consider renaming for clarity


### **education**

* 8 values; **university.degree** most frequent
* Ordinal variable → may require ranking
* Very low count for *illiterate* (18 records)
* Contains formatting issues (`basic.4y`, `high.school`)
* Missing (`unknown`) can reasonably be filled with mode: **university.degree**



### **default**

* Highly imbalanced:

  * no: 32,588
  * unknown: 8,597
  * yes: 3
* **Analytically unreliable** (likely underreported)
*  Recommendation: Exclude from main analysis and document as data quality issue



### **housing & loan**

* Both contain exactly **990 “unknown” values**
* Suggests a **shared group of non-respondents**
* Not an error — reflects real customer behavior



### **month**

* Only **10 months present** (Jan & Feb missing)
* Distribution:

  * May dominates (13,769 contacts)
  * December lowest (182)
*  Important: Do NOT treat as full-year data
* Indicates possible **seasonal campaign strategy**



### **poutcome**

* 86.3% = **“nonexistent”** (no prior contact)
* This is a **valid category**, not missing data
* Customers with previous **“success” (1,373)** are highly valuable targets


### **y (Target Variable)**

* no: 36,548
* yes: 4,640
* **Conversion rate: 11.3%**
* Strong class imbalance (~9:1)
* Recommendation: Use **percentages/KPIs**, not raw counts, in visuals



### Key Takeaways

* Data is largely realistic but contains **formatting issues and hidden missing values**
* Some variables (e.g., `default`) are **not analytically useful**
* Clear **class imbalance** and **seasonal trends**
* Strong opportunity for **customer targeting and campaign optimization analysis**



In [135]:
# Confirm pdays=999 dominance, duration range, campaign outliers, and age range.

print(df[['age','campaign','pdays',
          'previous','duration']].describe())
print()
print("pdays=999 count:", (df['pdays'] == 999).sum())
print("duration=0 count:", (df['duration'] == 0).sum())
print("campaign max:", df['campaign'].max())

               age      campaign         pdays      previous      duration
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000
mean      40.02406      2.567593    962.475454      0.172963    258.285010
std       10.42125      2.770014    186.910907      0.494901    259.279249
min       17.00000      1.000000      0.000000      0.000000      0.000000
25%       32.00000      1.000000    999.000000      0.000000    102.000000
50%       38.00000      2.000000    999.000000      0.000000    180.000000
75%       47.00000      3.000000    999.000000      0.000000    319.000000
max       98.00000     56.000000    999.000000      7.000000   4918.000000

pdays=999 count: 39673
duration=0 count: 4
campaign max: 56


The numerical features reveal several critical data quality and analytical considerations. The pdays variable is dominated by the placeholder value 999, appearing in 96.3% of rows, which renders its statistical measures meaningless; this value must be replaced with NaN before any analysis. Only 3.7% of customers were previously contacted, forming a small but highly valuable returning-customer segment. The duration column should be dropped entirely, as it represents post-call information not available at prediction time, includes invalid cases (duration = 0), and contains extreme outliers (up to 4,918 seconds) that would heavily skew analysis. The campaign variable shows that most customers were contacted 1–3 times, but extreme outliers (up to 56 contacts) indicate aggressive over-contacting and should be capped in visualizations to avoid distortion. In contrast, age is clean and well-distributed, ranging from 17 to 98 with a mean of 40, making it suitable for segmentation without additional cleaning.

In [136]:
before = len(df)
df = df.drop_duplicates().reset_index(drop=True)
after = len(df)
print(f"Rows removed: {before - after}")
print(f"Rows remaining: {after:,}")

Rows removed: 12
Rows remaining: 41,176


In [137]:
df.drop(columns=['duration'], inplace=True)
print("duration column dropped.")
print("Remaining columns:", df.shape[1])

duration column dropped.
Remaining columns: 20


In [138]:
# Only replace "unknown" in the columns where it is a missing placeholder. Do NOT touch poutcome — "nonexistent" there is valid documented dat

unknown_cols = ['job','marital','education',
                'default','housing','loan']

for col in unknown_cols:
    count = (df[col] == 'unknown').sum()
    df[col] = df[col].replace('unknown', np.nan)
    print(f"{col}: {count} unknowns → NaN")

print()
print("NaN counts after replacement:")
print(df[unknown_cols].isnull().sum())

job: 330 unknowns → NaN
marital: 80 unknowns → NaN
education: 1730 unknowns → NaN
default: 8596 unknowns → NaN
housing: 990 unknowns → NaN
loan: 990 unknowns → NaN

NaN counts after replacement:
job           330
marital        80
education    1730
default      8596
housing       990
loan          990
dtype: int64


In [139]:
# Under 5% missing → fill with mode (most frequent). Over 5% missing → label as "Unknown" so Power BI can show it transparently.

threshold = 0.05

for col in unknown_cols:
    pct = df[col].isnull().mean()
    if pct < threshold:
        fill_val = df[col].mode()[0]
        df[col] = df[col].fillna(fill_val)      # fixed
        print(f"{col}: {pct*100:.1f}% missing → filled with '{fill_val}'")
    else:
        df[col] = df[col].fillna('Unknown')     # fixed
        print(f"{col}: {pct*100:.1f}% missing → kept as 'Unknown'")

job: 0.8% missing → filled with 'admin.'
marital: 0.2% missing → filled with 'married'
education: 4.2% missing → filled with 'university.degree'
default: 20.9% missing → kept as 'Unknown'
housing: 2.4% missing → filled with 'yes'
loan: 2.4% missing → filled with 'no'


In [140]:
df['pdays'] = df['pdays'].replace(999, np.nan)

df['previously_contacted'] = df['pdays'].notna().map(
    {True: 'Yes', False: 'No'}
)

print("pdays NaN count:", df['pdays'].isnull().sum())
print(df['previously_contacted'].value_counts())

pdays NaN count: 39661
previously_contacted
No     39661
Yes     1515
Name: count, dtype: int64


In [141]:
#### Standardise text values "Admin." and "admin." as different categories and often cause confusion on our code or when using it on powerbi best is Strip and lowercase everything.

In [142]:
# Identify all columns with 'object' (string) data types
str_cols = df.select_dtypes(include=['object']).columns

for col in str_cols:
    df[col] = df[col].str.strip().str.lower()

# Remove trailing dot from "admin." specifically
df['job'] = df['job'].str.replace(r'\.$', '', regex=True)

print(df['job'].unique())

['housemaid' 'services' 'admin' 'blue-collar' 'technician' 'retired'
 'management' 'unemployed' 'self-employed' 'entrepreneur' 'student']


In [143]:
# Officially documented as a directive: "divorced" includes widowed. Rename for accurate Power BI reporting.
df['marital'] = df['marital'].replace(
    'divorced', 'divorced/widowed'
)
print(df['marital'].value_counts())

marital
married             25001
single              11564
divorced/widowed     4611
Name: count, dtype: int64


In [144]:
# Replace dot notation('basic.4y') with readable labels for Power BI axis and tooltips.
edu_labels = {
    'basic.4y':            'Basic (4 years)',
    'basic.6y':            'Basic (6 years)',
    'basic.9y':            'Basic (9 years)',
    'high.school':         'High School',
    'professional.course': 'Professional Course',
    'university.degree':   'University Degree',
    'illiterate':          'Illiterate',
    'unknown':             'Unknown'
}
df['education'] = df['education'].replace(edu_labels)
print(df['education'].value_counts())

education
University Degree      13894
High School             9512
Basic (9 years)         6045
Professional Course     5240
Basic (4 years)         4176
Basic (6 years)         2291
Illiterate                18
Name: count, dtype: int64


In [145]:
# Education is ordinal — assign a numeric rank so Power BI sorts it correctly on charts instead of alphabetically.
edu_order = {
    'illiterate':          0,
    'Basic (4 years)':     1,
    'Basic (6 years)':     2,
    'Basic (9 years)':     3,
    'High School':         4,
    'Professional Course': 5,
    'University Degree':   6,
    'Unknown':             7
}
df['education_rank'] = df['education'].map(edu_order)
print(df[['education','education_rank']].drop_duplicates()
        .sort_values('education_rank'))

                education  education_rank
0         Basic (4 years)             1.0
3         Basic (6 years)             2.0
5         Basic (9 years)             3.0
1             High School             4.0
6     Professional Course             5.0
7       University Degree             6.0
5392           Illiterate             NaN


In [146]:
# Age as raw integers makes poor Power BI charts. Grouped bands make clean bar charts and slicers.

df['age_band'] = pd.cut(
    df['age'],
    bins=[0, 25, 35, 45, 55, 65, 100],
    labels=['Under 25','25–34','35–44',
            '45–54','55–64','65+']
)
print(df['age_band'].value_counts().sort_index())

age_band
Under 25     1665
25–34       14844
35–44       12839
45–54        8247
55–64        2963
65+           618
Name: count, dtype: int64


In [147]:
# Without sorting columns, Power BI will sort months alphabetically (apr, aug, dec…) instead of chronologically (jan, feb, mar…).
month_order = {'jan':1,'feb':2,'mar':3,'apr':4,
               'may':5,'jun':6,'jul':7,'aug':8,
               'sep':9,'oct':10,'nov':11,'dec':12}

day_order = {'mon':1,'tue':2,'wed':3,
             'thu':4,'fri':5}

df['month_order']   = df['month'].map(month_order)
df['day_order']     = df['day_of_week'].map(day_order)

print(df[['month','month_order']].drop_duplicates()
        .sort_values('month_order'))
print()
print(df[['day_of_week','day_order']].drop_duplicates()
        .sort_values('day_order'))

      month  month_order
27682   mar            3
27964   apr            4
0       may            5
7762    jun            6
12136   jul            7
18817   aug            8
37876   sep            9
23990   oct           10
24057   nov           11
27672   dec           12

     day_of_week  day_order
0            mon          1
360          tue          2
756          wed          3
1146         thu          4
1578         fri          5


In [148]:
# Keep the text column for Power BI labels. Add a numeric 0/1 column for KPI cards, conversion rate calculations, and any future modelling.

df['subscribed'] = df['y'].map({'yes': 1, 'no': 0})

print(df['subscribed'].value_counts())
print(f"\nConversion rate: {df['subscribed'].mean()*100:.1f}%")

subscribed
0    36537
1     4639
Name: count, dtype: int64

Conversion rate: 11.3%


In [149]:
# Dots in column names break DAX formulas in Power BI silently. Rename all four economic columns now.
df.rename(columns={
    'emp.var.rate':   'emp_var_rate',
    'cons.price.idx': 'cons_price_idx',
    'cons.conf.idx':  'cons_conf_idx',
    'nr.employed':    'nr_employed'
}, inplace=True)

print("Renamed columns:")
print([c for c in df.columns if '_' in c])

Renamed columns:
['day_of_week', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'nr_employed', 'previously_contacted', 'education_rank', 'age_band', 'month_order', 'day_order']


In [150]:
### Validate The Clean Data

In [151]:
# Only pdays should have nulls remaining — everything else should be zero.
nulls = df.isnull().sum()
print("Columns with nulls:")
print(nulls[nulls > 0])
print()
print("Total null cells:", df.isnull().sum().sum())

Columns with nulls:
pdays             39661
education_rank       18
dtype: int64

Total null cells: 39679


In [152]:
# Confirm row count is close to original (41,188) and column count is correct.
print(f"Final shape:   {df.shape}")
print(f"Rows kept:     {len(df):,} of 41,188 original")
print(f"Columns:       {df.shape[1]}")
print(f"\nAll columns:\n{list(df.columns)}")

Final shape:   (41176, 26)
Rows kept:     41,176 of 41,188 original
Columns:       26

All columns:
['age', 'job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'campaign', 'pdays', 'previous', 'poutcome', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed', 'y', 'previously_contacted', 'education_rank', 'age_band', 'month_order', 'day_order', 'subscribed']


In [153]:
# Confirm every engineered column has clean, expected values.
print("=== age_band ===")
print(df['age_band'].value_counts().sort_index())

print("\n=== previously_contacted ===")
print(df['previously_contacted'].value_counts())

print("\n=== education (readable labels) ===")
print(df['education'].value_counts())

print("\n=== marital ===")
print(df['marital'].value_counts())

print("\n=== subscribed ===")
print(df['subscribed'].value_counts())

=== age_band ===
age_band
Under 25     1665
25–34       14844
35–44       12839
45–54        8247
55–64        2963
65+           618
Name: count, dtype: int64

=== previously_contacted ===
previously_contacted
no     39661
yes     1515
Name: count, dtype: int64

=== education (readable labels) ===
education
University Degree      13894
High School             9512
Basic (9 years)         6045
Professional Course     5240
Basic (4 years)         4176
Basic (6 years)         2291
Illiterate                18
Name: count, dtype: int64

=== marital ===
marital
married             25001
single              11564
divorced/widowed     4611
Name: count, dtype: int64

=== subscribed ===
subscribed
0    36537
1     4639
Name: count, dtype: int64


In [154]:
print(df.columns.tolist())

['age', 'job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'campaign', 'pdays', 'previous', 'poutcome', 'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m', 'nr_employed', 'y', 'previously_contacted', 'education_rank', 'age_band', 'month_order', 'day_order', 'subscribed']


In [155]:
# The full health report on the clean dataset — your final quality gate before export.
df.info()
print()
df.head(3)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41176 entries, 0 to 41175
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   age                   41176 non-null  int64   
 1   job                   41176 non-null  object  
 2   marital               41176 non-null  object  
 3   education             41176 non-null  object  
 4   default               41176 non-null  object  
 5   housing               41176 non-null  object  
 6   loan                  41176 non-null  object  
 7   contact               41176 non-null  object  
 8   month                 41176 non-null  object  
 9   day_of_week           41176 non-null  object  
 10  campaign              41176 non-null  int64   
 11  pdays                 1515 non-null   float64 
 12  previous              41176 non-null  int64   
 13  poutcome              41176 non-null  object  
 14  emp_var_rate          41176 non-null  float64 
 15  co

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,cons_conf_idx,euribor3m,nr_employed,y,previously_contacted,education_rank,age_band,month_order,day_order,subscribed
0,56,housemaid,married,Basic (4 years),no,no,no,telephone,may,mon,...,-36.4,4.857,5191.0,no,no,1.0,55–64,5,1,0
1,57,services,married,High School,unknown,no,no,telephone,may,mon,...,-36.4,4.857,5191.0,no,no,4.0,55–64,5,1,0
2,37,services,married,High School,no,yes,no,telephone,may,mon,...,-36.4,4.857,5191.0,no,no,4.0,35–44,5,1,0


In [156]:
# Save clean dataset

df.to_csv('bank_marketing_clean.csv', index=False)
print("Saved: bank_marketing_clean.csv")
print(f"Rows: {len(df):,}  |  Columns: {df.shape[1]}")

Saved: bank_marketing_clean.csv
Rows: 41,176  |  Columns: 26


In [157]:
### Documentation of our Data cleaning process

In [158]:
print("=" * 40)
print("   CLEANING SUMMARY")
print("=" * 40)
print(f"  Original rows:    41,188")
print(f"  Final rows:       {len(df):,}")
print(f"  Rows removed:     {41188 - len(df)}")
print(f"  Final columns:    {df.shape[1]}")
print(f"  Null cells left:  {df.isnull().sum().sum()}")
print(f"  Conversion rate:  {df['subscribed'].mean()*100:.1f}%")
print("=" * 40)
print("\n  New columns added:")
new_cols = ['previously_contacted','age_band','subscribed',
            'education_rank','month_order','day_order']
for c in new_cols:
    print(f"    + {c}")
print("\n  Columns dropped:")
print("    - duration  (data leakage)")
print("\n  Columns renamed:")
print("    emp.var.rate   → emp_var_rate")
print("    cons.price.idx → cons_price_idx")
print("    cons.conf.idx  → cons_conf_idx")
print("    nr.employed    → nr_employed")
print("=" * 40)

   CLEANING SUMMARY
  Original rows:    41,188
  Final rows:       41,176
  Rows removed:     12
  Final columns:    26
  Null cells left:  39679
  Conversion rate:  11.3%

  New columns added:
    + previously_contacted
    + age_band
    + subscribed
    + education_rank
    + month_order
    + day_order

  Columns dropped:
    - duration  (data leakage)

  Columns renamed:
    emp.var.rate   → emp_var_rate
    cons.price.idx → cons_price_idx
    cons.conf.idx  → cons_conf_idx
    nr.employed    → nr_employed


In [159]:
### To import and chech where your clean dataset is saved

In [160]:
import os

print(os.getcwd()) 
print(os.path.exists('bank_marketing_clean.csv'))

C:\Users\HP\Pandas Project\Data Cleaning
True
